test/Fencer/Rules/Test.hs

```diff
@@ -71,6 +81,87 @@ test_loadRulesRecursively =
         (sortOn domainDefinitionId [domain1, domain2])
         (sortOn domainDefinitionId definitions)
 
+-- | Test that a rule limit unit change adds a new counter and leaves
+-- the old one intact.
+--
+-- TODO(md)-2019-10-11: Sometimes this test non-deterministically fails with:
+--
+-- Got wrong gRPC error response
+--         expected: ClientIOError (GRPCIOBadStatusCode StatusUnknown
+--           (StatusDetails {unStatusDetails = "rate limit descriptor
+--             list must not be empty"}))
+--          but got: ClientIOError (GRPCIOBadStatusCode StatusUnavailable
+--            (StatusDetails {unStatusDetails = "Endpoint read failed"}))
+test_ruleLimitUnitChange :: TestTree
+test_ruleLimitUnitChange =
+  -- TODO(md): creating a server here sometimes clashes with executing
+  -- server tests concurrently, which occasionally leads to test
+  -- failures. Fix this either by making sure the port hasn't been
+  -- binded or in some other way.
```

Comment:
The simplest way probably would be to create a global `IORef` (using `unsafePerformIO`) storing the number of a free port and access it from within the tests. Alternatively, we can package the tests together using

```haskell
-- Manual locking to prevent race conditions in tests that run concurrently.
testConcurrentGroup :: String -> [TestTree] -> TestTree
testConcurrentGroup groupName trees =
    withResource (newMVar ()) mempty $ \getLock ->
        let testSingle tree =
                withResource
                    (getLock >>= takeMVar)
                    (\_ -> getLock >>= flip putMVar ())
                    (\_ -> tree)
            in testGroup groupName $ map testSingle trees
```

but it probably doesn't play well with

```
{-# OPTIONS_GHC -F -pgmF tasty-discover -optF --tree-display #-}
```

(which I dislike anyway).

Another option would be to catch exceptions arising during creation of the server, but I'd say it's too hacky.

Either way, I'd fix it before merging the PR.

Reply:
Thanks for the comments! I realized that starting with Tast 1.2 there is a way to declare dependencies between tests. I used that to avoid running both server and rule tests at the same time.

Reply:
Oh, interesting, I wasn't aware of that. I'd make a small note about why we need this `after AllFinish`, otherwise +1.